src/rest.rs

```diff
@@ -924,22 +923,40 @@ impl Format {
     }
 }
 
-pub const DEFAULT_FORMAT: Format = Format::MessagePack;
+pub struct DecodeRaw<T>(PhantomData<T>);
 
-/// A pagination item handler which decodes each message.
-#[derive(Clone)]
-pub struct MessageItemHandler {
-    opts: Option<ChannelOptions>,
+pub trait Decode {
+    type Data: Clone + Send;
+    type Item: DeserializeOwned + Send + 'static;
+    fn decode(item: &mut Self::Item, data: &Self::Data);
 }
 
-impl MessageItemHandler {
-    pub fn new(opts: Option<ChannelOptions>) -> Self {
-        Self { opts }
+impl Decode for Message {
+    type Data = Option<ChannelOptions>;
+    type Item = Self;
+
+    fn decode(item: &mut Self::Item, data: &Self::Data) {
+        crate::rest::decode(&mut item.data, &mut item.encoding, data.as_ref());
     }
 }
 
-impl<T: Decode> http::PaginatedItemHandler<T> for MessageItemHandler {
-    fn handle(&self, msg: &mut T) {
-        msg.decode(self.opts.as_ref());
+impl Decode for Stats {
+    type Data = ();
+    type Item = Self;
+    fn decode(_item: &mut Self::Item, _data: &Self::Data) {}
+}
+
+impl Decode for PresenceMessage {
+    type Data = Option<ChannelOptions>;
```

Comment:
The associated type is named `Data`, then here we set it to `ChannelOptions`, and around the code we call it `handler`, which I found a little confusing.

Perhaps we should name the associated type `Options` (i.e. options for performing the decode), and then have:

```rust
type Options = Option<ChannelOptions>;
```

and:

```rust
paginated_request_with_options(... opts: T::Options)
```

?

Reply:
Options may confuse with std::option::Option<T>. The name could be different though I guess, and handler changed to data/whatever we change data to yeah.

Reply:
Changed handler to data for consistency. When we decide on a name we can then change all data to whatever.

Reply:
Couldn't think of a better name so went with Options